provider/pkg/schema/extension_resource.go

```diff
@@ -0,0 +1,272 @@
+package schema
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/mattbaird/jsonpatch"
+	"github.com/pulumi/pulumi-aws-native/provider/pkg/client"
+	"github.com/pulumi/pulumi-go-provider/resourcex"
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/pkg/v3/resource/provider"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/diag"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/wI2L/jsondiff"
+)
+
+type ExtensionResource struct {
+	Type         string
+	Properties   map[string]any
+	WriteOnly    []string
+	TagsProperty string
+	TagsStyle    TagsStyle
+}
+
+func (e *ExtensionResource) ToSpec() CloudAPIResource {
+	return CloudAPIResource{
+		CfType:       e.Type,
+		WriteOnly:    e.WriteOnly,
+		TagsProperty: e.TagsProperty,
+		TagsStyle:    e.TagsStyle,
+	}
+}
+
+func (e *ExtensionResource) ToOutputs(outputs map[string]any) resource.PropertyMap {
+	return resource.NewPropertyMapFromMap(map[string]any{
+		"type":         e.Type,
+		"properties":   e.Properties,
+		"writeOnly":    e.WriteOnly,
+		"tagsProperty": e.TagsProperty,
+		"tagsStyle":    e.TagsStyle,
+		"outputs":      outputs,
+	})
+}
+
+type extensionResource struct {
+	host   *provider.HostClient
+	client client.CloudControlClient
+}
+
+func NewExtensionResource(host *provider.HostClient, client client.CloudControlClient) *extensionResource {
+	return &extensionResource{
+		host:   host,
+		client: client,
+	}
+}
+
+func (r *extensionResource) Create(ctx context.Context, urn resource.URN, inputs resource.PropertyMap) (identifier *string, outputs resource.PropertyMap, err error) {
```

Comment:
I'm missing unit tests around everything in this file. With client mocking, it should be very doable, and very much warranted for the logic complexity-wise.

Reply:
Sure, that would be nice to have at some point.

Reply:
This PR adds 500 lines of net-new code. I consider tests to be a requirement in this case, not a nice-to-have.

Our default mode shall be ~full coverage of all new or meaningfully modified code, and every exception must be motivated.